### PR TITLE
 Allow axes with arbitrary bounds and positions.

### PR DESCRIPTION
**Pull request by elijahsummers**
Makes axes more configurable. These aspects are exposed now:
* axes bounds
* axes component position
* whether axes are considered fixed (ticks will be limited to the bounds provided) or not (ticks may expand beyond existing bounds)
* whether axes bounds influence the plot bounds
* which TickRenderer is used.

```scala
val points = Seq(Point(1, 1), Point(1.5, 1.1), Point(2.5,1.5), Point(2.9, 2.5), Point(3, 3))
val filler = Seq("Lorem", "ipsum", "dolor", "sit", "amet", "consectetur")
LinePlot(points)
  // Note discrete axes are still "banded/boxed" such that ticks don't point to values, but to the center of a
  // band for that value
  .discreteAxis(Seq("foo", "bar", "baz"), Seq(1d, 2, 10), Position.Bottom, updatePlotBounds = false)
  .discreteAxis(filler, filler.indices.map(_.toDouble), Position.Right, updatePlotBounds = false)
  .continuousAxis(plot => plot.xbounds, Position.Top, tickRenderer = Some(TickRenderer.AxisTickRenderer(
    Position.Top,
    rotateText = 315
  )))
  .continuousAxis(_ => Bounds(0, 100000), Position.Left, updatePlotBounds = false)
  .xGrid().yGrid()
  .frame()
  .render(Extent(400, 300))
```

screen shot 2018-07-12 at 9 06 48 am

### CODE REVIEW

File: shared/src/main/scala/com/cibo/evilplot/plot/components/Axes.scala [hunk 1]
```diff
@@ -213,30 +217,132 @@ object Axes {
...
+      * @param fixedBounds      Force ticks to match bounds.
+      */
+    def continuousAxis(
+      boundsFn: Plot => Bounds,
+      position: Position,
+      tickCount: Option[Int] = None,
+      tickRenderer: Option[TickRenderer] = None,
+      labelFormatter: Option[Double => String] = None,
+      tickCountRange: Option[Seq[Int]] = None,
+      updatePlotBounds: Boolean = true,
```
Comment (zacharygraziano) [inline shared/src/main/scala/com/cibo/evilplot/plot/components/Axes.scala]: This stuff always confuses me, so I want to clarify: is `updatePlotBounds` setting the plot's transformation to match the axis bounds or the axis label bounds?

File: shared/src/main/scala/com/cibo/evilplot/plot/renderers/TickRenderer.scala [hunk 2]
```diff
@@ -48,49 +49,79 @@ object TickRenderer {
...
+    Position.Left,
+    length,
+    thickness,
+    0
+  )
+
+  //TODO revisit alignment for rotated text labels so that tick is aligned to "highest" part of text
+  //     e.g. left-side for 1-89 & 181 - 269, right-side for 91 - 179 & 271-359
+  //     or expose alignment?
+  def AxisTickRenderer(
```
Comment (zacharygraziano) [inline shared/src/main/scala/com/cibo/evilplot/plot/renderers/TickRenderer.scala]: can this be `axisTickRenderer` and can you add Scaladoc?

File: shared/src/main/scala/com/cibo/evilplot/plot/components/Axes.scala [hunk 3]
```diff
@@ -213,30 +217,132 @@ object Axes {
...
+      fixedBounds: Boolean = true
+    )(implicit theme: Theme): Plot = {
+      val defaultRotation = if (position == Position.Bottom || position == Position.Top) {
+        theme.elements.categoricalXAxisLabelOrientation
```
Comment (zacharygraziano) [inline shared/src/main/scala/com/cibo/evilplot/plot/components/Axes.scala]: This is the continuous function but these theme values are for categorical axes.

File: shared/src/main/scala/com/cibo/evilplot/plot/components/Axes.scala (hunk 1 above)
Comment (elijahsummers) [reply shared/src/main/scala/com/cibo/evilplot/plot/components/Axes.scala]: It doesn't affect the transformation (except I suppose indirectly through setting the plot bounds). If `updatePlotBounds` is true, then the existing behavior is used which updates the plot bounds to match the axis bounds (AxisDescriptor#axisBounds).
```scala
// new implementation of this behavior
component +: plot.ybounds(component.getDescriptor(plot, fixedBounds).axisBounds)
// old replaced version
//component +: plot.ybounds(component.getDescriptor(plot, plot.yfixed).axisBounds)
```
If `updatePlotBounds` is false, then the axis component is added to the plot without touching the plot bounds at all.
```scala
component +: plot
```

**Pull request merged**